Mathlib/Tactic/Bound/Attribute.lean

```diff
@@ -0,0 +1,111 @@
+/-
+Copyright (c) 2024 Geoffrey Irving. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Geoffrey Irving
+-/
+import Mathlib.Init.ZeroOne
+import Mathlib.Tactic.Bound.Init
+import Qq
+
+/-!
+# The `bound` attribute
+
+Any lemma tagged with `@[bound]` is registered as an apply rule for the `bound` tactic, by
+converting it to either `norm apply` or `safe apply <priority>`.  The classification is based
+on the number and types of the lemmas hypotheses.
+-/
+
+open Lean (MetaM)
+open Qq
+
+namespace Bound
+
+initialize Lean.registerTraceClass `bound.attribute
+
+variable {u : Lean.Level} {α : Q(Type u)}
+
+/-- Check if an expression is zero -/
+def isZero (e : Q($α)) : MetaM Bool :=
+  match e with
+  | ~q(@OfNat.ofNat.{u} _ 0 $i) => return true
+  | ~q(@Zero.zero.{u} _ $i) => return true
+  | _ => return false
+
+/-- Map the arguments of an inequality expression to a score -/
+def ineqPriority (a b : Q($α)) : MetaM ℕ := do
+  return if (← isZero a) || (← isZero b) then 1 else 10
+
+/-- Map a hypothesis type to a score -/
+partial def hypPriority (hyp : Q(Prop)) : MetaM ℕ := do
+  match hyp with
+    -- Conjections add scores
+    | ~q($a ∧ $b) => pure <| (← hypPriority a) + (← hypPriority b)
+    -- Guessing (disjunction) gets a big penalty
+    | ~q($a ∨ $b) => pure <| 100 + (← hypPriority a) + (← hypPriority b)
+    -- Inequalities get score 1 if they contain zero, 10 otherwise
+    | ~q(@LE.le _ $i $a $b) => ineqPriority a b
+    | ~q(@LT.lt _ $i $a $b) => ineqPriority a b
+    | ~q(@GE.ge _ $i $b $a) => ineqPriority a b
+    | ~q(@GT.gt _ $i $b $a) => ineqPriority a b
+    -- Assume anything else is non-relevant
+    | _ => pure 0
+
+/-- Map a type to a score -/
+def typePriority (decl : Lean.Name) (type : Lean.Expr) : MetaM ℕ :=
+  Lean.Meta.forallTelescope type fun xs t ↦ do
+    checkResult t
+    xs.foldlM (fun (t : ℕ) x ↦ do return t + (← argPriority x)) 0
+  where
+  /-- Score the type of argument `x` -/
+  argPriority (x : Lean.Expr) : MetaM ℕ := do
+    hypPriority (← Lean.Meta.inferType x)
+  /-- Insist that our conclusion is an inequality -/
+  checkResult (t : Q(Prop)) : MetaM Unit := do match t with
+    | ~q(@LE.le _ $i $a $b) => return ()
+    | ~q(@LT.lt _ $i $a $b) => return ()
+    | ~q(@GE.ge _ $i $b $a) => return ()
+    | ~q(@GT.gt _ $i $b $a) => return ()
+    | _ => throwError (f!"`{decl}` has invalid type `{type}` as a bound lemma")
```

Comment:
Have you tried this using `match_expr` rather than just `Qq`? I'd prefer, if it makes no difference, to rely on simpler / better maintained technology.

Reply:
Someone told me to use `Qq` while I was developing this, but I'm fine switching if that's desirable. I haven't heard of `match_expr`, though: where's it documented?

Reply:
Hmm, I looked into `match_expr` a bit, and if possible I'd like to stick with `Qq` for at least this first commit. `match_expr` doesn't seem to have support for literal zeros, is used nowhere else in `mathlib4`, and has no documentation. The first match I tried to convert was `~q(@OfNat.ofNat.{u} _ 0 $i)`, but using an explicit `0` was a syntax error for `match_expr`, and I can't find any literal matching in the `lean4` repo. I'd be happy to do a later rewrite from `Qq` to `match_expr` if it gets more support, but it seems thin at the moment.